src/Ormolu/Parser.hs

```diff
@@ -173,11 +181,23 @@ normalizeModule hsmod =
       IEGroup _ _ s -> isBlankDocString s
       IEDoc _ s -> isBlankDocString s
       _ -> False
-
     dropBlankTypeHaddocks = \case
       L _ (HsDocTy _ ty s) :: LHsType GhcPs
         | isBlankDocString s -> ty
       a -> a
+    enforceParensAroundConstraints :: HsType GhcPs -> HsType GhcPs
+    enforceParensAroundConstraints = \case
+      HsQualTy pass lqs t -> HsQualTy pass (patchContext lqs) t
+      x -> x
+    enforceParensInClassConstraints :: TyClDecl GhcPs -> TyClDecl GhcPs
+    enforceParensInClassConstraints = \case
+      ClassDecl {..} -> ClassDecl {tcdCtxt = patchContext <$> tcdCtxt, ..}
+      x -> x
+    patchContext :: LHsContext GhcPs -> LHsContext GhcPs
+    patchContext (L l qs) = L l $ case qs of
+      [x@(L _ (HsParTy _ _))] -> [x]
+      [x@(L lx _)] -> [L lx (HsParTy EpAnnNotUsed x)]
+      xs -> xs
```

Comment:
Hmm, what about directly matching for `LHsContext` instead of only matching for a fixed list of types wherein it is contained? This would have the advantage of making it impossible to miss cases. The only caveat might be that due to `LHsContext` just being a type synonym for a located list of types, but if GHC were to start using a located list of types for something else, they would probably want different EPAs for them (as there can only be one [`Anno (LHsContext pass)`](https://hackage.haskell.org/package/ghc-9.4.4/docs/Language-Haskell-Syntax-Extension.html#t:Anno), so they would create newtypes.

Concretely, this would catch constraints in existential types and GADTs: Even with this PR, this is a fixed point:
```haskell
data Foo = forall a. Show a => Foo a

data Bar where
  Bar :: Show a => a -> Bar
```
With e.g. https://gist.github.com/amesgen/9ebd0bd7cc03689e6013c1f926a20345, parens are added, and at least the test suite passes.

Reply:
I also thought of matching on `LHsContext`, however it being a type synonym for a list of located types (a thing that seemed too common to be safe to modify) dissuaded me from following that path. I do agree with your reasoning regarding the `Anno` type family. Good one!